### PR TITLE
[CORE-823] Implement queries for x/ratelimit; get limiter and capacity together in keeper

### PR DESCRIPTION
**Pull request by teddyding**
### Changelist
- Implemented queries for `x/ratelimite` module
- Refactored keeper logic to query corresponding `capacitiy` and `limiter` together, using `GetLimiterCapacityListForDenom`

### Test Plan
Unit tests and query integration tests

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

### CODE REVIEW

File: protocol/x/ratelimit/util/capacity.go [hunk 1]
```diff
@@ -28,33 +27,22 @@ import (
...
 	timeSinceLastBlock time.Duration,
 ) (
 	newCapacityList []dtypes.SerializableInt,
-	err error,
 ) {
 	// Declare new capacity list to be populated.
-	newCapacityList = make([]dtypes.SerializableInt, len(prevCapacityList))
+	newCapacityList = make([]dtypes.SerializableInt, len(limiterCapacityList))
 
-	if len(limitParams.Limiters) != len(prevCapacityList) {
```
Comment (teddyding) [inline protocol/x/ratelimit/util/capacity.go:41]: This function now takes in list of `LimiterCapacity`, each of which is a tuple, so no need to throw error here

File: protocol/x/ratelimit/client/cli/query_capacity_by_denom_test.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		[]string{
+			fmt.Sprintf("--%s=json", tmcli.OutputFlag),
+			assettypes.AssetUsdc.Denom,
+		})
+
+	require.NoError(t, err)
+	var resp types.QueryCapacityByDenomResponse
+	require.NoError(t, net.Config.Codec.UnmarshalJSON(out.Bytes(), &resp))
+	require.Equal(t,
+		[]types.LimiterCapacity{
```
Comment (tqin7) [inline protocol/x/ratelimit/client/cli/query_capacity_by_denom_test.go:38]: For my understanding, is this expected list of `LimiterCapacity` from default genesis?

File: protocol/x/ratelimit/module_test.go [hunk 3]
```diff
@@ -107,7 +128,10 @@ func TestAppModuleBasic_GetTxCmd(t *testing.T) {
 }
 
 func TestAppModuleBasic_GetQueryCmd(t *testing.T) {
-	createAppModuleBasic(t)
+	am := createAppModuleBasic(t)
 
-	// TODO(CORE-823): implement query for `x/ratelimit`
+	cmd := am.GetQueryCmd()
+	require.Equal(t, "vest", cmd.Use)
```
Comment (tqin7) [inline protocol/x/ratelimit/module_test.go]: nit:
```suggestion
	require.Equal(t, "ratelimit", cmd.Use)
```
below assertions need to be updated accordingly too

File: protocol/x/ratelimit/keeper/keeper.go [hunk 4]
```diff
@@ -244,47 +244,80 @@ func (k Keeper) UpdateAllCapacitiesEndBlocker(
...
 	limitParams := k.GetAllLimitParams(ctx)
 	for _, limitParams := range limitParams {
-		k.updateCapacityForLimitParams(ctx, limitParams, timeSinceLastBlock)
+		k.updateCapacityForDenom(ctx, limitParams.Denom, timeSinceLastBlock)
 	}
 }
 
 // updateCapacityForLimitParams calculates current baseline for a denom and recovers some amount of capacity
 // towards baseline.
 // Assumes that the `LimitParams` exist in state.
```
Comment (tqin7) [inline protocol/x/ratelimit/keeper/keeper.go:253]: nit: perhaps don't need line 253 anymore?

**Review by tqin7 — APPROVED** [review]

File: protocol/x/ratelimit/keeper/grpc_query.go [hunk 5]
```diff
@@ -11,16 +12,38 @@ import (
...
-	ctx context.Context,
+	goCtx context.Context,
 	req *types.ListLimitParamsRequest,
 ) (*types.ListLimitParamsResponse, error) {
-	// 	// TODO(CORE-823): implement query for `x/ratelimit`
-	return nil, status.Errorf(codes.Unimplemented, "method ListLimitParams not implemented")
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
```
Comment (BrendanChou) [inline protocol/x/ratelimit/keeper/grpc_query.go:21]: looks like `sdkCtx := sdk.UnwrapSDKContext(ctx)` is more canonical variable naming in the codebase?

**Review by BrendanChou — APPROVED** [review]

File: protocol/x/ratelimit/client/cli/query_capacity_by_denom_test.go (hunk 2 above)
Comment (teddyding) [reply protocol/x/ratelimit/client/cli/query_capacity_by_denom_test.go:38]: Yes, added comment to clarify.

**Pull request merged**